File: services/api/src/topicpilot_api/topic_lifecycle_contract.py
"""Frozen Topic Lifecycle V0 contract vocabulary and availability boundary.

This module records the existing V2 contract recovered by the WS1 preflight.
It does not approve the provisional numeric policy or production publication.
"""

from __future__ import annotations

from typing import Literal

OWNER_LIFECYCLE_STAGES = ("萌芽", "發酵", "主升", "成熟", "衰退")
BACKEND_LIFECYCLE_STAGES = ("SPROUTING", "FERMENTING", "MAIN_RISE", "MATURE", "DECLINING")
BACKEND_TO_OWNER_LIFECYCLE_STAGE = dict(
    zip(BACKEND_LIFECYCLE_STAGES, OWNER_LIFECYCLE_STAGES, strict=True)
)

# These aliases are retained presentation lineage, not additional stages.
LEGACY_PRESENTATION_ALIASES = {"高檔整理": "成熟", "退潮": "衰退"}
ADJACENT_NON_LIFECYCLE_STATES = frozenset({"升溫", "降溫", "WARMING", "COOLING", "觀察"})

LifecycleAvailability = Literal[
    "AVAILABLE",
    "FORMAL_AVAILABLE",
    "SHADOW_AVAILABLE",
    "INSUFFICIENT_DATA",
    "PENDING",
    "PREVIEW",
    "NOT_AVAILABLE",
    "WAITING_FOR_FORMAL_LINEAGE",
    "FAIL_CLOSED",
]

LIFECYCLE_AVAILABILITY_STATES = (
    "AVAILABLE",
    "FORMAL_AVAILABLE",
    "SHADOW_AVAILABLE",
    "INSUFFICIENT_DATA",
    "PENDING",
    "PREVIEW",
    "NOT_AVAILABLE",
    "WAITING_FOR_FORMAL_LINEAGE",
    "FAIL_CLOSED",
)


def is_backend_lifecycle_stage(value: str | None) -> bool:
    return value in BACKEND_LIFECYCLE_STAGES


__all__ = [
    "ADJACENT_NON_LIFECYCLE_STATES",
    "BACKEND_LIFECYCLE_STAGES",
    "BACKEND_TO_OWNER_LIFECYCLE_STAGE",
    "LEGACY_PRESENTATION_ALIASES",
    "LIFECYCLE_AVAILABILITY_STATES",
    "OWNER_LIFECYCLE_STAGES",
    "LifecycleAvailability",
    "is_backend_lifecycle_stage",
]

File: apps/web/app/lib/topic-lifecycle-contract.ts
/** Existing Topic Lifecycle V0 vocabulary; this is not a sixth-stage design. */
export const OWNER_LIFECYCLE_STAGES = ["萌芽", "發酵", "主升", "成熟", "衰退"] as const;

export type OwnerLifecycleStage = typeof OWNER_LIFECYCLE_STAGES[number];

export const BACKEND_TO_OWNER_LIFECYCLE_STAGE: Record<string, OwnerLifecycleStage> = {
  SPROUTING: "萌芽",
  FERMENTING: "發酵",
  MAIN_RISE: "主升",
  MATURE: "成熟",
  DECLINING: "衰退",
};

// Retained presentation lineage; these values are never accepted from the
// formal backend stage mapper as a sixth or replacement stage.
export const LEGACY_PRESENTATION_ALIASES = {
  "高檔整理": "成熟",
  "退潮": "衰退",
} as const;

export const LIFECYCLE_AVAILABILITY_STATES = [
  "AVAILABLE",
  "FORMAL_AVAILABLE",
  "SHADOW_AVAILABLE",
  "INSUFFICIENT_DATA",
  "PENDING",
  "PREVIEW",
  "NOT_AVAILABLE",
  "WAITING_FOR_FORMAL_LINEAGE",
  "FAIL_CLOSED",
] as const;

export type LifecycleAvailability = typeof LIFECYCLE_AVAILABILITY_STATES[number];

export function ownerStageFromBackend(stage: string | null | undefined): OwnerLifecycleStage | null {
  return stage ? BACKEND_TO_OWNER_LIFECYCLE_STAGE[stage] ?? null : null;
}
